File: sns_rooster/lib/models/employee.dart
class Employee {
  final String id;
  final String userId;
  final String firstName;
  final String lastName;
  final String email;
  final String employeeId;
  final DateTime hireDate;
  final String? position;
  final String? department;

  Employee({
    required this.id,
    required this.userId,
    required this.firstName,
    required this.lastName,
    required this.email,
    required this.employeeId,
    required this.hireDate,
    this.position,
    this.department,
  });

  factory Employee.fromJson(Map<String, dynamic> json) {
    return Employee(
      id: json['_id'] as String,
      userId: json['userId'] as String,
      firstName: json['firstName'] as String,
      lastName: json['lastName'] as String,
      email: json['email'] as String,
      employeeId: json['employeeId'] as String,
      hireDate: DateTime.parse(json['hireDate'] as String),
      position: json['position'] as String?,
      department: json['department'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'userId': userId,
      'firstName': firstName,
      'lastName': lastName,
      'email': email,
      'employeeId': employeeId,
      'hireDate': hireDate.toIso8601String(),
      'position': position,
      'department': department,
    };
  }

  String get name => '$firstName $lastName';

  @override
  String toString() {
    return 'Employee(id: $id, firstName: $firstName, lastName: $lastName, email: $email, employeeId: $employeeId, hireDate: $hireDate, position: $position, department: $department)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is Employee &&
        other.id == id &&
        other.firstName == firstName &&
        other.lastName == lastName &&
        other.email == email &&
        other.employeeId == employeeId &&
        other.hireDate == hireDate &&
        other.position == position &&
        other.department == department;
  }

  @override
  int get hashCode {
    return id.hashCode ^
        firstName.hashCode ^
        lastName.hashCode ^
        email.hashCode ^
        employeeId.hashCode ^
        hireDate.hashCode ^
        position.hashCode ^
        department.hashCode;
  }

  Employee copyWith({
    String? id,
    String? userId,
    String? firstName,
    String? lastName,
    String? email,
    String? employeeId,
    DateTime? hireDate,
    String? position,
    String? department,
  }) {
    return Employee(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      firstName: firstName ?? this.firstName,
      lastName: lastName ?? this.lastName,
      email: email ?? this.email,
      employeeId: employeeId ?? this.employeeId,
      hireDate: hireDate ?? this.hireDate,
      position: position ?? this.position,
      department: department ?? this.department,
    );
  }
}

File: rooster-backend/models/Employee.js
const mongoose = require('mongoose');

const employeeSchema = new mongoose.Schema({
  // Company Association
  companyId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Company',
    required: true,
    index: true
  },
  firstName: {
    type: String,
    required: true,
    trim: true
  },
  lastName: {
    type: String,
    required: true,
    trim: true
  },
  email: {
    type: String,
    required: true,
    trim: true,
    lowercase: true
  },
  employeeId: {
    type: String,
    required: true,
    trim: true
  },
  hireDate: {
    type: Date,
    default: Date.now
  },
  position: {
    type: String,
    trim: true
  },
  department: {
    type: String,
    trim: true
  },
  hourlyRate: {
    type: Number,
    default: 0,
    min: 0,
  },
  monthlySalary: {
    type: Number,
    default: 0,
    min: 0,
  },
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    unique: true,
  },
  isActive: {
    type: Boolean,
    default: true
  },
  employeeType: {
    type: String,
    enum: ['Permanent', 'Temporary'],
    default: 'Permanent',
  },
  employeeSubType: {
    type: String,
    enum: ['Full-time', 'Part-time', 'Casual', null],
    default: null,
  },
  
  // Enterprise Features - Multi-Location Support
  locationId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Location',
    index: true
  },
  
  // Enterprise Features - Performance Management
  performanceLevel: {
    type: String,
    enum: ['entry', 'junior', 'intermediate', 'senior', 'expert', 'lead'],
    default: 'entry'
  },
  lastPerformanceReview: {
    type: Date
  },
  nextPerformanceReview: {
    type: Date
  },
  
  // Enterprise Features - Training Management
  trainingHistory: [{
    trainingId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Training'
    },
    status: {
      type: String,
      enum: ['assigned', 'enrolled', 'in_progress', 'completed', 'failed', 'dropped'],
      default: 'assigned'
    },
    enrolledAt: Date,
    completedAt: Date,
    score: Number,
    certificate: {
      filename: String,
      url: String,
      issuedAt: Date
    }
  }],
  
  // Enterprise Features - Expense Management
  expenseLimit: {
    type: Number,
    min: 0,
    default: 0
  },
  expenseApprover: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  
  // Additional employee fields
  manager: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  emergencyContact: {
    name: String,
    relationship: String,
    phone: String,
    email: String
  },
  skills: [{
    type: String,
    trim: true
  }],
  certifications: [{
    name: String,
    issuer: String,
    issueDate: Date,
    expiryDate: Date,
    certificateUrl: String
  }],
  notes: {
    type: String,
    trim: true
  }
});

// Compound unique indexes for company-specific uniqueness
employeeSchema.index({ companyId: 1, email: 1 }, { unique: true });
employeeSchema.index({ companyId: 1, employeeId: 1 }, { unique: true });

const Employee = mongoose.model('Employee', employeeSchema);

module.exports = Employee; 